src/types/mod.rs

```diff
@@ -150,11 +150,13 @@ macro_rules! pyobject_native_static_type_object(
 #[doc(hidden)]
 #[macro_export]
 macro_rules! pyobject_native_type_info(
-    ($name:ty, $typeobject:expr, $module:expr $(, #checkfunction=$checkfunction:path)? $(;$generics:ident)*) => {
+    ($name:ty, $typeobject:expr, $module:expr, $python_name:expr $(, #checkfunction=$checkfunction:path)? $(;$generics:ident)*) => {
         // SAFETY: macro caller has upheld the safety contracts
         unsafe impl<$($generics,)*> $crate::type_object::PyTypeInfo for $name {
             const NAME: &'static str = stringify!($name);
-            const MODULE: ::std::option::Option<&'static str> = $module;
+            const MODULE: ::std::option::Option<&'static str> = ::std::option::Option::Some($module);
+            #[cfg(feature = "experimental-inspect")]
+            const TYPE_HINT: $crate::inspect::TypeHint = $crate::inspect::TypeHint::module_attr($module, $python_name);
```

Comment:
I'm not totally sure we want to change and/or populate `NAME` and `MODULE`, given the deprecation. Given the eventual result once we remove those is to just fill `TYPE_HINT`, how about instead an adjustment to the macro so that `NAME` and `MODULE` are left as-is, and the macros just get an additional (optional?) argument which is the type hint (passed as a real `TypeHint` value?).

(I say optional because that'll be easier for downstream crates using these macros, i.e. `rust-numpy`, which will otherwise have awkward interactions with the optional feature here.)

Reply:
I have added back the old `#module` parameter to not touch `MODULE`

>passed as a real TypeHint value

This would require an `experimental-inspect` gate leading to duplicated macro calls in all usage places. It looks very painful to me.

I have kept the two new parameters module + name required, it makes call places looks imho nicer than having something like `#type_hint = ("collections.abc", "Iterable")` and I think we want to ensure all users of this macro provide type hint information. These macros are not part of pyO3 public API and migration does not looks too bad to me. But glad to make this optional if you think it would be better (just answer "yes please")